core/src/main/scala/org/apache/gearpump/cluster/master/AppManager.scala

```diff
@@ -0,0 +1,256 @@
+import org.apache.gearpump.cluster.MasterToAppMaster._
+import org.apache.gearpump.cluster.MasterToClient.{ReplayApplicationResult, ResolveAppIdResult, ShutdownApplicationResult, SubmitApplicationResult}
+import org.apache.gearpump.cluster.WorkerToAppMaster._
+import org.apache.gearpump.cluster._
+import org.apache.gearpump.cluster.scheduler.{Resource, ResourceAllocation, ResourceRequest}
+import org.apache.gearpump.transport.HostPort
+import org.apache.gearpump.util.ActorSystemBooter._
+import org.apache.gearpump.util.Constants._
+import org.apache.gearpump.util._
+import org.slf4j.Logger
+
+import scala.collection.JavaConverters._
+import scala.concurrent.Future
+import scala.concurrent.duration.Duration
+import scala.util.{Failure, Success}
+
+/**
+ * AppManager is dedicated part of Master to manager applicaitons
+ */
+private[cluster] class AppManager(masterHA : ActorRef, kvService: ActorRef, launcher: AppMasterLauncherFactory) extends Actor with Stash with TimeOutScheduler{
+  private val LOG: Logger = LogUtil.getLogger(getClass)
+
+  private val executorId : Int = APPMASTER_DEFAULT_EXECUTOR_ID
+  private val systemconfig = context.system.settings.config
+
+  implicit val timeout = FUTURE_TIMEOUT
+  implicit val executionContext = context.dispatcher
+
+  //next available appId
+  private var appId: Int = 0
+
+  //from appid to appMaster data
+  private var appMasterRegistry = Map.empty[Int, (ActorRef, AppMasterRuntimeInfo)]
+
+  def receive: Receive = null
+
+  (masterHA ? GetMasterState).asInstanceOf[Future[GetMasterStateResult]].map(self ! _)
+  context.become(waitForMasterState)
+
+  def waitForMasterState: Receive = {
+    case MasterState(state) =>
+      val set = state.map(_.appId)
+      if (set.isEmpty) {
+        this.appId = 0
+      } else {
+        this.appId = state.map(_.appId).max + 1
+      }
+      context.become(receiveHandler)
+      unstashAll()
+    case GetMasterStateFailed(ex) =>
+      LOG.error("Failed to get master state, shutting down master to avoid data corruption...")
+      context.parent ! PoisonPill
+    case msg =>
+      LOG.info(s"Get message ${msg.getClass.getSimpleName}")
+      stash()
+  }
+
+  def receiveHandler : Receive = {
+    val msg = "Application Manager started. Ready for application submission..."
+    System.out.println(msg)
+    LOG.info(msg)
+
+    clientMsgHandler orElse appMasterMessage orElse selfMsgHandler orElse workerMessage orElse appDataStoreService orElse terminationWatch
+  }
+
+  def clientMsgHandler: Receive = {
+    case SubmitApplication(app, jar, username) =>
+      LOG.info(s"AppManager Submiting Application $appId...")
+      val client = sender
+      val appLauncher = context.actorOf(launcher.props(appId, executorId, app, jar, username, context.parent, Some(client)), s"launcher${appId}_${Util.randInt}")
+
+      val appState = new ApplicationState(appId, 0, app, jar, username, null)
+      masterHA ! UpdateMasterState(appState)
+      appId += 1
```

Comment:
Thanks for finding this.

the code to reply to client is part of AppMasterLauncher.

I have just add the UT in AppMasterLauncherSpec.